### PR TITLE
Create Proper Migration tests

### PR DESCRIPTION
**Pull request by zeripath**
Unfortunately our testing regime has so far meant that migrations do not
get proper testing.

This PR begins the process of creating migration tests for this.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
TODO:

- [ ] wire into makefile
- [ ] wire into .drone.yml

**Comment by zeripath** [comment]
OK it's wired in..

The structure is a bit of a hack with the current inferior migration tests still being run from integrations/migration-test - these should probably be either moved in to migrations as proper tests or just removed once we have enough tests to be happy that they are no longer needed.

File: models/migrations/v177_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	defer deferable()
+
+	var issueLabels []*IssueLabel
+	preMigration := map[int64]*IssueLabel{}
+	postMigration := map[int64]*IssueLabel{}
+
+	if err := x.Sync2(new(IssueLabel)); err != nil {
+		assert.NoError(t, err)
+		return
+	}
```
Comment (zeripath) [inline models/migrations/v177_test.go:48]:
```suggestion
```

Comment (zeripath) [inline models/migrations/v177_test.go]:
```suggestion
```

File: models/migrations/migrations_test.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	x.SetConnMaxLifetime(setting.Database.ConnMaxLifetime)
+	return x, nil
+}
+
+func prepareTestEnv(t *testing.T, skip int, syncModels ...interface{}) (*xorm.Engine, func()) {
+	t.Helper()
+	ourSkip := 2
+	ourSkip += skip
+	deferFn := PrintCurrentTest(t, ourSkip)
+	// assert.NoError(t, models.LoadFixtures())
```
Comment (zeripath) [inline models/migrations/migrations_test.go]:
```suggestion
```

File: models/migrations/migrations_test.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	// so use log file to instead print to stdout.
+	x.SetLogger(models.NewXORMLogger(setting.Database.LogSQL))
+	x.ShowSQL(setting.Database.LogSQL)
+	x.SetMaxOpenConns(setting.Database.MaxOpenConns)
+	x.SetMaxIdleConns(setting.Database.MaxIdleConns)
+	x.SetConnMaxLifetime(setting.Database.ConnMaxLifetime)
+	return x, nil
+}
+
+func prepareTestEnv(t *testing.T, skip int, syncModels ...interface{}) (*xorm.Engine, func()) {
```
Comment (zeripath) [inline models/migrations/migrations_test.go:198]:
```suggestion
// prepareTestEnv prepares the test environment. The skip parameter should usually be 0. Provide models to be sync'd
// with the database - in particular any models you expect fixtures to be loaded from.
//
// fixtures in `models/migrations/fixtures/<TestName>` will be loaded automatically
func prepareTestEnv(t *testing.T, skip int, syncModels ...interface{}) (*xorm.Engine, func()) {
```

File: models/migrations/migrations_test.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+			return x, deferFn
+		}
+		if err := models.LoadFixtures(x); err != nil {
+			t.Errorf("error whilst loading fixtures from %s: %v", fixturesDir, err)
+			return x, deferFn
+		}
+	} else if !os.IsNotExist(err) {
+		t.Errorf("unexpected error whilst checking for existence of fixtures: %v", err)
+	} else {
+		t.Logf("no fixtures found for: %s", fixturesDir)
```
Comment (zeripath) [inline models/migrations/migrations_test.go]:
```suggestion
		t.Logf("no fixtures found in: %s", fixturesDir)
```

**Comment by zeripath** [comment]
Oh that's interesting there's a bug with drop table columns on mssql ...

**Review by 6543 — APPROVED** [review]

**Review by kdumontnu — APPROVED** [review]
Thanks for adding these! 💯 

Do you think it would be good practice to add a comment block at the top of `v176_test.go` and `v177_test.go` to explain at high level what the migrations intend to do? Ex. "create index X", etc.

**Comment by zeripath** [comment]
I think such comments would probably be better on the migration functions themselves.

**Review by 6543 — APPROVED** [review]

**Pull request merged**